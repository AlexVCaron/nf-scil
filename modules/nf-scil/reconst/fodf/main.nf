
process RECONST_FODF {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_1.6.0.sif':
        'scilus/scilus:1.6.0' }"

    input:
        tuple val(meta), path(dwi), path(bval), path(bvec), path(mask), path(frf)

    output:
        tuple val(meta), path("*fodf.nii.gz")           , emit: fodf
        path "versions.yml"                             , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    def dwi_shell_tolerance = task.ext.dwi_shell_tolerance ? "--tolerance " + task.ext.dwi_shell_tolerance : ""
    def fodf_shells = task.ext.fodf_shells ?: "\$(cut -d ' ' --output-delimiter=\$'\\n' -f 1- $bval | awk -F' ' '{v=int(\$1)}{if(v>=$task.ext.min_fodf_shell_value|| v<=$task.ext.b0_thr_extract_b0)print v}' | uniq)"
    def sh_order = task.ext.sh_order ? "--sh_order " + task.ext.sh_order : ""
    def sh_basis = task.ext.sh_basis ? "--sh_basis " + task.ext.sh_basis : ""
    def processes = task.ext.processes ? "--processes " + task.ext.processes : ""
    def set_mask = mask ? "--mask $mask" : ""

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    scil_extract_dwi_shell.py $dwi $bval $bvec $fodf_shells \
        dwi_fodf_shells.nii.gz bval_fodf_shells bvec_fodf_shells \
        $dwi_shell_tolerance -f

    scil_compute_ssst_fodf.py dwi_fodf_shells.nii.gz bval_fodf_shells bvec_fodf_shells $frf ${prefix}__fodf.nii.gz \
        $sh_order $sh_basis --force_b0_threshold \
        $set_mask $processes

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 1.6.0
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    scil_extract_dwi_shell.py -h
    scil_compute_ssst_fodf.py -h

    touch ${prefix}__fodf.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 1.6.0
    END_VERSIONS
    """
}
