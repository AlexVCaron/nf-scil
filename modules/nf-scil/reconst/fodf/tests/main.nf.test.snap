{
    "reconst - fodf_msmt": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__csf_fodf.nii.gz:md5,44ff13caaa26130249895f780ebd0a75",
                            "test__fodf.nii.gz:md5,ed63fb06bd1736a205ff9ac833defbe8",
                            "test__gm_fodf.nii.gz:md5,29a2c27ac120480036924663c381079d",
                            "test__wm_fodf.nii.gz:md5,ed63fb06bd1736a205ff9ac833defbe8"
                        ]
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__wm_fodf.nii.gz:md5,ed63fb06bd1736a205ff9ac833defbe8"
                    ]
                ],
                "10": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_total.nii.gz:md5,96b64b378f6621f0173a01b7e3d6f36b"
                    ]
                ],
                "11": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_sum.nii.gz:md5,614363bda38e66e07b0c154bb2909735"
                    ]
                ],
                "12": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__nufo.nii.gz:md5,79b4d04795e5ddb2f7e4ece0731f95c8"
                    ]
                ],
                "13": [
                    
                ],
                "14": [
                    "versions.yml:md5,8f5a73278d63aac3bf8073c5d0ef488b"
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__gm_fodf.nii.gz:md5,29a2c27ac120480036924663c381079d"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__csf_fodf.nii.gz:md5,44ff13caaa26130249895f780ebd0a75"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__vf.nii.gz:md5,4bcef849fa12d9c5919653d41ec8ba53"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__vf_rgb.nii.gz:md5,e7dc8b93171f71d0c8c770b3b9713344"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peaks.nii.gz:md5,c3bdb59ba45ff454c01141554e475207"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_values.nii.gz:md5,e5e32f5a28d756f178c1471be636a3f4"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_indices.nii.gz:md5,483c40a518a86cfb3b9a71cf0ce626b6"
                    ]
                ],
                "9": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_max.nii.gz:md5,aa740712d22ee08c4029afa7a506d3bf"
                    ]
                ],
                "afd_max": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_max.nii.gz:md5,aa740712d22ee08c4029afa7a506d3bf"
                    ]
                ],
                "afd_sum": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_sum.nii.gz:md5,614363bda38e66e07b0c154bb2909735"
                    ]
                ],
                "afd_total": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_total.nii.gz:md5,96b64b378f6621f0173a01b7e3d6f36b"
                    ]
                ],
                "csf_fodf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__csf_fodf.nii.gz:md5,44ff13caaa26130249895f780ebd0a75"
                    ]
                ],
                "fodf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__csf_fodf.nii.gz:md5,44ff13caaa26130249895f780ebd0a75",
                            "test__fodf.nii.gz:md5,ed63fb06bd1736a205ff9ac833defbe8",
                            "test__gm_fodf.nii.gz:md5,29a2c27ac120480036924663c381079d",
                            "test__wm_fodf.nii.gz:md5,ed63fb06bd1736a205ff9ac833defbe8"
                        ]
                    ]
                ],
                "gm_fodf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__gm_fodf.nii.gz:md5,29a2c27ac120480036924663c381079d"
                    ]
                ],
                "nufo": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__nufo.nii.gz:md5,79b4d04795e5ddb2f7e4ece0731f95c8"
                    ]
                ],
                "peak_indices": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_indices.nii.gz:md5,483c40a518a86cfb3b9a71cf0ce626b6"
                    ]
                ],
                "peak_values": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_values.nii.gz:md5,e5e32f5a28d756f178c1471be636a3f4"
                    ]
                ],
                "peaks": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peaks.nii.gz:md5,c3bdb59ba45ff454c01141554e475207"
                    ]
                ],
                "vent_mask": [
                    
                ],
                "versions": [
                    "versions.yml:md5,8f5a73278d63aac3bf8073c5d0ef488b"
                ],
                "vf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__vf.nii.gz:md5,4bcef849fa12d9c5919653d41ec8ba53"
                    ]
                ],
                "vf_rgb": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__vf_rgb.nii.gz:md5,e7dc8b93171f71d0c8c770b3b9713344"
                    ]
                ],
                "wm_fodf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__wm_fodf.nii.gz:md5,ed63fb06bd1736a205ff9ac833defbe8"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-07T14:37:15.07704"
    },
    "reconst - fodf": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__fodf.nii.gz:md5,c45f0391a462d358930809502dd2da15"
                    ]
                ],
                "1": [
                    
                ],
                "10": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_total.nii.gz:md5,9bac465d58b8662c3dce2c4937300a5a"
                    ]
                ],
                "11": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_sum.nii.gz:md5,a849e3eb8c67cc176e88daca01669eef"
                    ]
                ],
                "12": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__nufo.nii.gz:md5,711f0ee8d01e71166099ffdf4c684b5f"
                    ]
                ],
                "13": [
                    
                ],
                "14": [
                    "versions.yml:md5,8f5a73278d63aac3bf8073c5d0ef488b"
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    
                ],
                "5": [
                    
                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peaks.nii.gz:md5,fbc8af94bce991a6ebe3bfb837d39a52"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_values.nii.gz:md5,4fd4a44c18544413c0e9bb672e60f709"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_indices.nii.gz:md5,d39ad843a7b86ac7042c44b87356e68b"
                    ]
                ],
                "9": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_max.nii.gz:md5,96b7a2dbab6306ca053b45bfad21be55"
                    ]
                ],
                "afd_max": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_max.nii.gz:md5,96b7a2dbab6306ca053b45bfad21be55"
                    ]
                ],
                "afd_sum": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_sum.nii.gz:md5,a849e3eb8c67cc176e88daca01669eef"
                    ]
                ],
                "afd_total": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_total.nii.gz:md5,9bac465d58b8662c3dce2c4937300a5a"
                    ]
                ],
                "csf_fodf": [
                    
                ],
                "fodf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__fodf.nii.gz:md5,c45f0391a462d358930809502dd2da15"
                    ]
                ],
                "gm_fodf": [
                    
                ],
                "nufo": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__nufo.nii.gz:md5,711f0ee8d01e71166099ffdf4c684b5f"
                    ]
                ],
                "peak_indices": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_indices.nii.gz:md5,d39ad843a7b86ac7042c44b87356e68b"
                    ]
                ],
                "peak_values": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_values.nii.gz:md5,4fd4a44c18544413c0e9bb672e60f709"
                    ]
                ],
                "peaks": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peaks.nii.gz:md5,fbc8af94bce991a6ebe3bfb837d39a52"
                    ]
                ],
                "vent_mask": [
                    
                ],
                "versions": [
                    "versions.yml:md5,8f5a73278d63aac3bf8073c5d0ef488b"
                ],
                "vf": [
                    
                ],
                "vf_rgb": [
                    
                ],
                "wm_fodf": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-01T19:21:03.487368"
    },
    "reconst - fodf_no_metrics": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__fodf.nii.gz:md5,c45f0391a462d358930809502dd2da15"
                    ]
                ],
                "1": [
                    
                ],
                "10": [
                    
                ],
                "11": [
                    
                ],
                "12": [
                    
                ],
                "13": [
                    
                ],
                "14": [
                    "versions.yml:md5,8f5a73278d63aac3bf8073c5d0ef488b"
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    
                ],
                "5": [
                    
                ],
                "6": [
                    
                ],
                "7": [
                    
                ],
                "8": [
                    
                ],
                "9": [
                    
                ],
                "afd_max": [
                    
                ],
                "afd_sum": [
                    
                ],
                "afd_total": [
                    
                ],
                "csf_fodf": [
                    
                ],
                "fodf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__fodf.nii.gz:md5,c45f0391a462d358930809502dd2da15"
                    ]
                ],
                "gm_fodf": [
                    
                ],
                "nufo": [
                    
                ],
                "peak_indices": [
                    
                ],
                "peak_values": [
                    
                ],
                "peaks": [
                    
                ],
                "vent_mask": [
                    
                ],
                "versions": [
                    "versions.yml:md5,8f5a73278d63aac3bf8073c5d0ef488b"
                ],
                "vf": [
                    
                ],
                "vf_rgb": [
                    
                ],
                "wm_fodf": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-01T20:22:29.171158"
    },
    "reconst - fodf_with_shells": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__fodf.nii.gz:md5,c45f0391a462d358930809502dd2da15"
                    ]
                ],
                "1": [
                    
                ],
                "10": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_total.nii.gz:md5,9bac465d58b8662c3dce2c4937300a5a"
                    ]
                ],
                "11": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_sum.nii.gz:md5,a849e3eb8c67cc176e88daca01669eef"
                    ]
                ],
                "12": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__nufo.nii.gz:md5,711f0ee8d01e71166099ffdf4c684b5f"
                    ]
                ],
                "13": [
                    
                ],
                "14": [
                    "versions.yml:md5,8f5a73278d63aac3bf8073c5d0ef488b"
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    
                ],
                "5": [
                    
                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peaks.nii.gz:md5,fbc8af94bce991a6ebe3bfb837d39a52"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_values.nii.gz:md5,4fd4a44c18544413c0e9bb672e60f709"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_indices.nii.gz:md5,d39ad843a7b86ac7042c44b87356e68b"
                    ]
                ],
                "9": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_max.nii.gz:md5,96b7a2dbab6306ca053b45bfad21be55"
                    ]
                ],
                "afd_max": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_max.nii.gz:md5,96b7a2dbab6306ca053b45bfad21be55"
                    ]
                ],
                "afd_sum": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_sum.nii.gz:md5,a849e3eb8c67cc176e88daca01669eef"
                    ]
                ],
                "afd_total": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_total.nii.gz:md5,9bac465d58b8662c3dce2c4937300a5a"
                    ]
                ],
                "csf_fodf": [
                    
                ],
                "fodf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__fodf.nii.gz:md5,c45f0391a462d358930809502dd2da15"
                    ]
                ],
                "gm_fodf": [
                    
                ],
                "nufo": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__nufo.nii.gz:md5,711f0ee8d01e71166099ffdf4c684b5f"
                    ]
                ],
                "peak_indices": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_indices.nii.gz:md5,d39ad843a7b86ac7042c44b87356e68b"
                    ]
                ],
                "peak_values": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_values.nii.gz:md5,4fd4a44c18544413c0e9bb672e60f709"
                    ]
                ],
                "peaks": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peaks.nii.gz:md5,fbc8af94bce991a6ebe3bfb837d39a52"
                    ]
                ],
                "vent_mask": [
                    
                ],
                "versions": [
                    "versions.yml:md5,8f5a73278d63aac3bf8073c5d0ef488b"
                ],
                "vf": [
                    
                ],
                "vf_rgb": [
                    
                ],
                "wm_fodf": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-01T20:38:04.916377"
    },
    "reconst - fodf_no_mask": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__fodf.nii.gz:md5,1c0aa14841e90acbdc048c6f0b117ce2"
                    ]
                ],
                "1": [
                    
                ],
                "10": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_total.nii.gz:md5,713fae8f03e15434034c21dc647f4a20"
                    ]
                ],
                "11": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_sum.nii.gz:md5,e907c21b54410ad69531b58b2c6a5e5d"
                    ]
                ],
                "12": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__nufo.nii.gz:md5,0e58d2571878aa13882f858cb4e572ad"
                    ]
                ],
                "13": [
                    
                ],
                "14": [
                    "versions.yml:md5,8f5a73278d63aac3bf8073c5d0ef488b"
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    
                ],
                "5": [
                    
                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peaks.nii.gz:md5,3919c8117b3b800f52fc7319f523eb13"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_values.nii.gz:md5,e92f556fe1a926d088590dd2a7d970d6"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_indices.nii.gz:md5,ab4c1129a3ede2db638115e71258a288"
                    ]
                ],
                "9": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_max.nii.gz:md5,db6d8bd847e08e2857173f572b46670a"
                    ]
                ],
                "afd_max": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_max.nii.gz:md5,db6d8bd847e08e2857173f572b46670a"
                    ]
                ],
                "afd_sum": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_sum.nii.gz:md5,e907c21b54410ad69531b58b2c6a5e5d"
                    ]
                ],
                "afd_total": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_total.nii.gz:md5,713fae8f03e15434034c21dc647f4a20"
                    ]
                ],
                "csf_fodf": [
                    
                ],
                "fodf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__fodf.nii.gz:md5,1c0aa14841e90acbdc048c6f0b117ce2"
                    ]
                ],
                "gm_fodf": [
                    
                ],
                "nufo": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__nufo.nii.gz:md5,0e58d2571878aa13882f858cb4e572ad"
                    ]
                ],
                "peak_indices": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_indices.nii.gz:md5,ab4c1129a3ede2db638115e71258a288"
                    ]
                ],
                "peak_values": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_values.nii.gz:md5,e92f556fe1a926d088590dd2a7d970d6"
                    ]
                ],
                "peaks": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peaks.nii.gz:md5,3919c8117b3b800f52fc7319f523eb13"
                    ]
                ],
                "vent_mask": [
                    
                ],
                "versions": [
                    "versions.yml:md5,8f5a73278d63aac3bf8073c5d0ef488b"
                ],
                "vf": [
                    
                ],
                "vf_rgb": [
                    
                ],
                "wm_fodf": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-01T19:50:32.073387"
    }
}